language-support/ts/daml-ledger/index.ts

```diff
@@ -329,6 +331,277 @@ type LedgerOptions = {
+    }
+
+    close(): void {
+        this.beforeClosing();
+        this.emit('close', { code: 4000, reason: "called .close()" });
+        this.removeAllListeners();
+    }
+
+}
+
+type QueryResponseStream<T extends object, K, I extends string> = StreamEventEmitter<T, K, I, readonly CreateEvent<T, K, I>[]>;
+
+type StreamingQuery<T extends object, K, I extends string> = {
+    template: Template<T, K, I>,
+    queries: Query<T>[],
+    stream: QueryResponseStream<T, K, I>,
+    offset?: string | null, // undefined -> nothing has been received, null -> the JSON API returned a null offset
+    state: Map<ContractId<T>, CreateEvent<T, K, I>>, // all JavaScript Map iterators preserve insertion order
+}
+
+type StreamingQueryRequest = {
+    templateIds: string[],
+    query?: object,
+    offset?: string,
+};
+
+function append<K, V>(map: Map<K, V[]>, key: K, value: V): void {
+    if (map.has(key)) {
+        map.get(key)!.push(value);
+    } else {
+        map.set(key, [value]);
+    }
+}
+
+/**
+ * @deprecated All usages of this function should be replaced by just
+ *             iterating over the iterator. For this to happen, the TS
+ *             compiler requires the --downlevelIteration flag, which
+ *             however does not play nicely with Jest when running the
+ *             tests.
+ */
+function materialize<A>(iterator: IterableIterator<A>): Array<A> {
+  return Array.from(iterator);
+}
+
+/**
+ * @internal
+ *
+ * A special handler for stream requests to the /v1/stream/query endpoint.
+ * The query endpoint supports providing offsets on a per-query basis.
+ * This class leverages this feature by multiplexing virtual streaming requests to a single web socket.
+ */
+class QueryStreamsManager {
+
+    private static readonly ENDPOINT: string = 'v1/stream/query';
+
+    // Mutable state BEGIN
+
+    // Ongoing streaming queries that will be the downstream consumers of web socket messages
+    private readonly queries: Set<StreamingQuery<object, unknown, string>> = new Set();
+
+    // Lookup tables used to route events to the relevant consumers:
+    //  - consumers for create events can be looked up based on their match index
+    //    - store the offset by which a match indexes needs to be shifted before the event is passed to the consumer
+    //  - archive events can be lookup up by template identifier
+    //    - this causes the consumer to observe what is known as "phantom archives", which are known and documented
+    private matchIndexLookupTable: [StreamingQuery<object, unknown, string>, number][] = [];
+    private templateIdsLookupTable: {[templateId: string]: Set<StreamingQuery<object, unknown, string>>} = {};
+
+    // Accumulates each query in a flattened form to be sent as a single request to the JSON API
+    private request: StreamingQueryRequest[] = [];
+
+    // web socket handle and associated properties
+    private ws: WebSocket | null = null;
+    private wsLiveSince: number | undefined = undefined;
```

Comment:
Why the distinction between null and undefined here?

Reply:
In my mind `undefined` is something which can be absent and will have an actual value at a later point, while `null` is something I explicitly want to define as such.

Would you have found

```suggestion
    private wsLiveSince?: number;
```

to be easier to read? And maybe use a symbol to denote when the web socket is shut down?

Reply:
As mentioned offline, going with `?` for both fields seems the most sensible thing to do.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/10221/commits/8e1889d1c686611adc7fa02238e26327df302e17